src/framework/net/webconnection.cpp

```diff
@@ -0,0 +1,290 @@
+
+    //Workaround for the abrupt termination of the websocket
+    g_dispatcher.addEvent([] {
+        g_game.forceLogout();
+    });
+}
+
+
+void WebConnection::connect(const std::string_view host, uint16_t port, const std::function<void()>& connectCallback, bool gameWorld)
+{
+    m_gameWorld = gameWorld;
+
+    m_connected = false;
+    m_connecting = true;
+    m_connectCallback = connectCallback;
+
+    std::string ip = std::string(host);
+    ip.length() == 0 ? ip = "localhost" : ip;
+
+#ifdef NDEBUG
+    const std::string prefix = "wss://";
+#else
+    const std::string prefix = "ws://";
+#endif
+
+    m_pthread = pthread_self();
+
+    const std::string url = prefix + ip + ":" + std::to_string(port);
+    EmscriptenWebSocketCreateAttributes attributes =
+    {
+        url.c_str(),
+        "binary",
+        EM_FALSE // if the webscocket should be created in the main thread. Currently not implemented by emscripten so this does nothing
+    };
+
+    m_websocket = emscripten_websocket_new(&attributes);
+
+    if (m_websocket < 1) {
+        if (m_errorCallback)
+            m_errorCallback(asio::error::network_unreachable);
+
+        close();
+        return;
+    }
+
+    emscripten_websocket_set_onopen_callback(m_websocket, this, ([](int /*eventType*/, const EmscriptenWebSocketOpenEvent* /*event*/, void* userData) -> EM_BOOL {
+        WebConnection* webConnection = static_cast<WebConnection*>(userData);
+        webConnection->m_connected = true;
+
+        if (webConnection->m_connectCallback) {
+            emscripten_dispatch_to_thread(webConnection->m_pthread, EM_FUNC_SIG_VI, reinterpret_cast<void*>(runOnConnectCallback), nullptr, &webConnection->m_connectCallback);
+        }
+
+        webConnection->m_connecting = false;
+
+        return EM_TRUE;
+    }));
+
+    emscripten_websocket_set_onerror_callback(m_websocket, this, ([](int /*eventType*/, const EmscriptenWebSocketErrorEvent* /*event*/, void* userData) -> EM_BOOL {
+        WebConnection* webConnection = static_cast<WebConnection*>(userData);
+        if (webConnection->m_errorCallback) {
+            emscripten_dispatch_to_thread(webConnection->m_pthread, EM_FUNC_SIG_VI, reinterpret_cast<void*>(runOnErrorCallback), nullptr, &webConnection->m_errorCallback);
+        }
+        if (webConnection->m_connected || webConnection->m_connecting) {
+            webConnection->close();
+        }
+        return EM_TRUE;
+    }));
+
+    emscripten_websocket_set_onclose_callback(m_websocket, this, ([](int /*eventType*/, const EmscriptenWebSocketCloseEvent* /*event*/, void* userData) -> EM_BOOL {
+        WebConnection* webConnection = static_cast<WebConnection*>(userData);
+        webConnection->close();
+        return EM_TRUE;
+    }));
+
+    emscripten_websocket_set_onmessage_callback(m_websocket, this, ([](int /*eventType*/, const EmscriptenWebSocketMessageEvent* webSocketEvent, void* userData) -> EM_BOOL {
+        uint32_t numBytes = webSocketEvent->numBytes;
```

Comment:
```suggestion
        auto numBytes = webSocketEvent->numBytes;
```
it's actually uint64_t: https://github.com/emscripten-core/emscripten/blob/10cb9d46cdd17e7a96de68137c9649d9a630fbc7/system/include/emscripten/fetch.h#L164C1-L165C1
but easier to just let the compiler deduce it with `auto`